File: src/data/repository/lastfmRepository.ts
// ═════════════════════════════════════════════════════════════════════
//  Last.fm Repository — Soundwave Music Platform
//
//  Provides cached, safe, read-only metadata & recommendation operations:
//  • Feature 1: Artist Bio & Wiki (artist.getInfo)
//  • Feature 2: Similar Songs & Similar Artists (track.getSimilar, artist.getSimilar)
//  • Feature 3: Album Backstory & Tracklist (album.getInfo)
//  • Feature 4: Global Top Charts (chart.getTopTracks, chart.getTopArtists)
//  • Feature 5: Genre & Mood Tags (tag.getTopTracks)
// ═════════════════════════════════════════════════════════════════════

import type { Artist } from '../models';
import {
  getLastfmArtist,
  getSimilarArtists,
  getSimilarTracks,
  getLastfmAlbumInfo,
  getLastfmTopArtists,
  getLastfmTopTracks,
  getLastfmTagTopTracks,
  getLastfmTrackInfo,
  getLastfmArtistTopTracks,
  type LastFmArtistInfo,
  type LastFmAlbumInfo,
  type SimilarTrack,
} from '../api/lastfmApi';

export class LastFmRepository {
  private static instance: LastFmRepository;

  private constructor() {}

  public static getInstance(): LastFmRepository {
    if (!LastFmRepository.instance) {
      LastFmRepository.instance = new LastFmRepository();
    }
    return LastFmRepository.instance;
  }

  /**
   * FEATURE 1: Fetches artist biography, global listener count, playcount, and genre tags.
   */
  async getArtistBioAndInfo(artistName: string): Promise<LastFmArtistInfo | null> {
    try {
      return await getLastfmArtist(artistName);
    } catch (e) {
      console.warn(`[LastFmRepository] getArtistBioAndInfo error for "${artistName}":`, e);
      return null;
    }
  }

  /**
   * FEATURE 2: Fetches musically and acoustically similar songs matching the seed song.
   */
  async getSimilarTracks(artist: string, trackTitle: string, limit = 10): Promise<SimilarTrack[]> {
    try {
      return await getSimilarTracks(artist, trackTitle, limit);
    } catch (e) {
      console.warn(`[LastFmRepository] getSimilarTracks error for "${trackTitle}":`, e);
      return [];
    }
  }

  /**
   * FEATURE 2 (cont): Fetches similar artists for related artist discovery.
   */
  async getSimilarArtists(artistName: string, limit = 8): Promise<Artist[]> {
    try {
      return await getSimilarArtists(artistName, limit);
    } catch (e) {
      console.warn(`[LastFmRepository] getSimilarArtists error for "${artistName}":`, e);
      return [];
    }
  }

  /**
   * FEATURE 3: Fetches album backstory/wiki summary, release date, listeners, and tracklist.
   */
  async getAlbumInfo(artist: string, albumTitle: string): Promise<LastFmAlbumInfo | null> {
    try {
      return await getLastfmAlbumInfo(artist, albumTitle);
    } catch (e) {
      console.warn(`[LastFmRepository] getAlbumInfo error for "${albumTitle}":`, e);
      return null;
    }
  }

  /**
   * FEATURE 4: Fetches real-time global top 50 tracks.
   */
  async getGlobalTopTracks(limit = 20): Promise<{ title: string; artist: string; image?: string; playcount?: number; listeners?: number }[]> {
    try {
      return await getLastfmTopTracks(limit);
    } catch (e) {
      console.warn('[LastFmRepository] getGlobalTopTracks error:', e);
      return [];
    }
  }

  /**
   * FEATURE 4 (cont): Fetches trending global artists.
   */
  async getGlobalTopArtists(limit = 20): Promise<Artist[]> {
    try {
      return await getLastfmTopArtists(limit);
    } catch (e) {
      console.warn('[LastFmRepository] getGlobalTopArtists error:', e);
      return [];
    }
  }

  /**
   * FEATURE 5: Fetches top-ranked songs for specific genre tags (e.g. 'bollywood', 'rock', 'pop', 'lofi', 'indie').
   */
  async getGenreTagTopTracks(tag: string, limit = 20): Promise<{ title: string; artist: string; image?: string; rank?: number }[]> {
    try {
      return await getLastfmTagTopTracks(tag, limit);
    } catch (e) {
      console.warn(`[LastFmRepository] getGenreTagTopTracks error for "${tag}":`, e);
      return [];
    }
  }

  /**
   * Helper: Fetches track backstory/wiki & tags.
   */
  async getTrackInfo(artist: string, track: string) {
    try {
      return await getLastfmTrackInfo(artist, track);
    } catch (e) {
      console.warn(`[LastFmRepository] getTrackInfo error for "${track}":`, e);
      return null;
    }
  }

  /**
   * Helper: Fetches top tracks for an artist.
   */
  async getArtistTopTracks(artistName: string, limit = 20) {
    try {
      return await getLastfmArtistTopTracks(artistName, limit);
    } catch (e) {
      console.warn(`[LastFmRepository] getArtistTopTracks error for "${artistName}":`, e);
      return [];
    }
  }
}

export const lastFmRepository = LastFmRepository.getInstance();

File: android/app/src/main/java/com/auramusic/app/data/lastfm/LastFmRepository.kt
package com.auramusic.app.data.lastfm

import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.util.concurrent.ConcurrentHashMap

/**
 * LastFmRepository (Kotlin)
 * Provides unauthenticated read-only Last.fm music metadata with safe runCatching
 * error handling and in-memory TTL caching.
 */
class LastFmRepository(
    private val apiKey: String = "b25b959554ed76058ac220b7b2e0a026",
    private val baseUrl: String = "https://ws.audioscrobbler.com/2.0/"
) {
    companion object {
        @Volatile
        private var instance: LastFmRepository? = null

        fun getInstance(apiKey: String = "b25b959554ed76058ac220b7b2e0a026"): LastFmRepository {
            return instance ?: synchronized(this) {
                instance ?: LastFmRepository(apiKey).also { instance = it }
            }
        }
    }

    private data class CacheEntry(val data: String, val timestamp: Long)
    private val cache = ConcurrentHashMap<String, CacheEntry>()
    private val cacheTtlMs = 15 * 60 * 1000L // 15 minutes

    private fun getFromCache(key: String): String? {
        val entry = cache[key] ?: return null
        if (System.currentTimeMillis() - entry.timestamp > cacheTtlMs) {
            cache.remove(key)
            return null
        }
        return entry.data
    }

    private fun saveToCache(key: String, data: String) {
        cache[key] = CacheEntry(data, System.currentTimeMillis())
    }

    private fun executeHttpGet(params: Map<String, String>): Result<String> = runCatching {
        val query = params.entries.joinToString("&") { (k, v) ->
            "${URLEncoder.encode(k, "UTF-8")}=${URLEncoder.encode(v, "UTF-8")}"
        }
        val fullUrl = "$baseUrl?$query&api_key=$apiKey&format=json&autocorrect=1"
        val cacheKey = fullUrl

        val cached = getFromCache(cacheKey)
        if (cached != null) {
            return@runCatching cached
        }

        val url = URL(fullUrl)
        val conn = url.openConnection() as HttpURLConnection
        conn.requestMethod = "GET"
        conn.connectTimeout = 8000
        conn.readTimeout = 8000
        conn.setRequestProperty("User-Agent", "Soundwave-Android/1.2")

        val responseCode = conn.responseCode
        if (responseCode != HttpURLConnection.HTTP_OK) {
            throw Exception("HTTP $responseCode from Last.fm API")
        }

        val reader = BufferedReader(InputStreamReader(conn.inputStream))
        val response = reader.use { it.readText() }
        conn.disconnect()

        saveToCache(cacheKey, response)
        response
    }

    /**
     * FEATURE 1: Artist Biography & Wiki (artist.getInfo)
     */
    fun getArtistInfo(artistName: String): Result<ArtistInfoResponse> = runCatching {
        val jsonStr = executeHttpGet(mapOf("method" to "artist.getInfo", "artist" to artistName)).getOrThrow()
        val root = JSONObject(jsonStr)
        val artistObj = root.optJSONObject("artist") ?: return@runCatching ArtistInfoResponse(null)

        val statsObj = artistObj.optJSONObject("stats")
        val bioObj = artistObj.optJSONObject("bio")
        val tagsObj = artistObj.optJSONObject("tags")

        val tagList = mutableListOf<LastFmTag>()
        tagsObj?.optJSONArray("tag")?.let { arr ->
            for (i in 0 until arr.length()) {
                val t = arr.optJSONObject(i)
                if (t != null) tagList.add(LastFmTag(name = t.optString("name", "")))
            }
        }

        ArtistInfoResponse(
            artist = LastFmArtist(
                name = artistObj.optString("name", artistName),
                mbid = artistObj.optString("mbid", null),
                url = artistObj.optString("url", null),
                stats = LastFmStats(
                    listeners = statsObj?.optString("listeners", "0"),
                    playcount = statsObj?.optString("playcount", "0")
                ),
                bio = LastFmBio(
                    summary = bioObj?.optString("summary", "")?.replace(Regex("<[^>]*>"), "")?.trim(),
                    content = bioObj?.optString("content", "")?.replace(Regex("<[^>]*>"), "")?.trim()
                ),
                tags = LastFmTags(tag = tagList)
            )
        )
    }

    /**
     * FEATURE 2: Similar Songs (track.getSimilar)
     */
    fun getSimilarTracks(artist: String, track: String, limit: Int = 10): Result<SimilarTracksResponse> = runCatching {
        val jsonStr = executeHttpGet(
            mapOf("method" to "track.getSimilar", "artist" to artist, "track" to track, "limit" to limit.toString())
        ).getOrThrow()
        val root = JSONObject(jsonStr)
        val tracksList = mutableListOf<SimilarTrackItem>()
        val arr = root.optJSONObject("similartracks")?.optJSONArray("track")
        if (arr != null) {
            for (i in 0 until arr.length()) {
                val item = arr.optJSONObject(i) ?: continue
                val artistObj = item.optJSONObject("artist")
                tracksList.add(
                    SimilarTrackItem(
                        name = item.optString("name", ""),
                        match = item.optString("match", "0"),
                        duration = item.optString("duration", "0"),
                        playcount = item.optString("playcount", "0"),
                        artist = LastFmArtistName(name = artistObj?.optString("name", "") ?: ""),
                        url = item.optString("url", null)
                    )
                )
            }
        }
        SimilarTracksResponse(similartracks = SimilarTracksList(track = tracksList))
    }

    /**
     * FEATURE 2 (cont): Similar Artists (artist.getSimilar)
     */
    fun getSimilarArtists(artist: String, limit: Int = 8): Result<SimilarArtistsResponse> = runCatching {
        val jsonStr = executeHttpGet(
            mapOf("method" to "artist.getSimilar", "artist" to artist, "limit" to limit.toString())
        ).getOrThrow()
        val root = JSONObject(jsonStr)
        val artistsList = mutableListOf<LastFmArtist>()
        val arr = root.optJSONObject("similarartists")?.optJSONArray("artist")
        if (arr != null) {
            for (i in 0 until arr.length()) {
                val item = arr.optJSONObject(i) ?: continue
                artistsList.add(
                    LastFmArtist(
                        name = item.optString("name", ""),
                        url = item.optString("url", null)
                    )
                )
            }
        }
        SimilarArtistsResponse(similarartists = SimilarArtistsList(artist = artistsList))
    }

    /**
     * FEATURE 3: Album Wiki & Backstory (album.getInfo)
     */
    fun getAlbumInfo(artist: String, album: String): Result<AlbumInfoResponse> = runCatching {
        val jsonStr = executeHttpGet(
            mapOf("method" to "album.getInfo", "artist" to artist, "album" to album)
        ).getOrThrow()
        val root = JSONObject(jsonStr)
        val alb = root.optJSONObject("album") ?: return@runCatching AlbumInfoResponse(null)

        val wikiObj = alb.optJSONObject("wiki")
        val tracksArr = alb.optJSONObject("tracks")?.optJSONArray("track")
        val trackList = mutableListOf<LastFmAlbumTrackItem>()
        if (tracksArr != null) {
            for (i in 0 until tracksArr.length()) {
                val t = tracksArr.optJSONObject(i) ?: continue
                trackList.add(
                    LastFmAlbumTrackItem(
                        name = t.optString("name", ""),
                        duration = t.optString("duration", "0"),
                        url = t.optString("url", null)
                    )
                )
            }
        }

        AlbumInfoResponse(
            album = LastFmAlbum(
                name = alb.optString("name", album),
                artist = alb.optString("artist", artist),
                listeners = alb.optString("listeners", "0"),
                playcount = alb.optString("playcount", "0"),
                wiki = LastFmBio(
                    summary = wikiObj?.optString("summary", "")?.replace(Regex("<[^>]*>"), "")?.trim(),
                    published = wikiObj?.optString("published", null)
                ),
                tracks = LastFmAlbumTracks(track = trackList),
                url = alb.optString("url", null)
            )
        )
    }

    /**
     * FEATURE 4: Global Top Charts (chart.getTopTracks & chart.getTopArtists)
     */
    fun getGlobalTopTracks(limit: Int = 20): Result<TopChartsTracksResponse> = runCatching {
        val jsonStr = executeHttpGet(
            mapOf("method" to "chart.getTopTracks", "limit" to limit.toString())
        ).getOrThrow()
        val root = JSONObject(jsonStr)
        val tracksList = mutableListOf<SimilarTrackItem>()
        val arr = root.optJSONObject("tracks")?.optJSONArray("track")
        if (arr != null) {
            for (i in 0 until arr.length()) {
                val item = arr.optJSONObject(i) ?: continue
                val artistObj = item.optJSONObject("artist")
                tracksList.add(
                    SimilarTrackItem(
                        name = item.optString("name", ""),
                        playcount = item.optString("playcount", "0"),
                        artist = LastFmArtistName(name = artistObj?.optString("name", "") ?: ""),
                        url = item.optString("url", null)
                    )
                )
            }
        }
        TopChartsTracksResponse(tracks = TopTracksList(track = tracksList))
    }

    fun getGlobalTopArtists(limit: Int = 20): Result<TopChartsArtistsResponse> = runCatching {
        val jsonStr = executeHttpGet(
            mapOf("method" to "chart.getTopArtists", "limit" to limit.toString())
        ).getOrThrow()
        val root = JSONObject(jsonStr)
        val artistsList = mutableListOf<LastFmArtist>()
        val arr = root.optJSONObject("artists")?.optJSONArray("artist")
        if (arr != null) {
            for (i in 0 until arr.length()) {
                val item = arr.optJSONObject(i) ?: continue
                artistsList.add(
                    LastFmArtist(
                        name = item.optString("name", ""),
                        stats = LastFmStats(listeners = item.optString("listeners", "0")),
                        url = item.optString("url", null)
                    )
                )
            }
        }
        TopChartsArtistsResponse(artists = TopArtistsList(artist = artistsList))
    }

    /**
     * FEATURE 5: Genre & Mood Tags (tag.getTopTracks)
     */
    fun getGenreTagTopTracks(tag: String, limit: Int = 20): Result<TagTopTracksResponse> = runCatching {
        val jsonStr = executeHttpGet(
            mapOf("method" to "tag.getTopTracks", "tag" to tag, "limit" to limit.toString())
        ).getOrThrow()
        val root = JSONObject(jsonStr)
        val tracksList = mutableListOf<SimilarTrackItem>()
        val arr = root.optJSONObject("tracks")?.optJSONArray("track")
        if (arr != null) {
            for (i in 0 until arr.length()) {
                val item = arr.optJSONObject(i) ?: continue
                val artistObj = item.optJSONObject("artist")
                tracksList.add(
                    SimilarTrackItem(
                        name = item.optString("name", ""),
                        artist = LastFmArtistName(name = artistObj?.optString("name", "") ?: ""),
                        url = item.optString("url", null)
                    )
                )
            }
        }
        TagTopTracksResponse(tracks = TopTracksList(track = tracksList))
    }
}
